docs/DifferentiableFunctions.md

```diff
@@ -0,0 +1,454 @@
+    func original(_ x: T) -> R {
+        return vjp(x).value
+    }
+    @differentiating(original)
+    func derivative(_ x: T) -> (value: R, pullback: (R.CotangentVector) -> T.CotangentVector) {
+        return vjp(x)
+    }
+    return original
+}
+```
+
+## `@differentiable` functions and automatic differentiation
+
+Automatic differentiation is the technique used by the compiler to automatically
+compute function derivatives. This document does not go into detail about
+automatic differentiation - but with an understanding of `@differentiable`
+functions and differentiation APIs, one can get a glimpse of how automatic
+differentiation works.
+
+The key differentiation API is the [`valueWithPullback`] function, which takes a
+`@differentiable` function and arguments and returns two things: the result of
+the function when applied to arguments, and a backpropagation function called a
+"pullback", which takes the gradient of the result and returns the gradient of
+the arguments.
+
+Let's consider the following function `foo`:
+
+```swift
+func foo(_ x: Float) -> Float {
+    let double = x + x
+    let result = double * double
+    return result
+}
+```
+
+Conceptually, here's how the compiler computes `valueWithPullback` for `foo`:
+
+```swift
+func fooValueWithPullback(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
+    // Replace function calls in `foo` with calls to `valueWithPullback`.
+    // Keep track of pullbacks and use them to compute pullback of `foo`.
+    let (double, doublePullback) =
+        valueWithPullback(at: x, x, in: (+) as @differentiable (Float, Float) -> Float)
+    let (result, resultPullback) =
+        valueWithPullback(at: double, double, in: (*) as @differentiable (Float, Float) -> Float)
+    let pullback: (Float) -> Float = { v in
+        let (𝛁result1, 𝛁result2) = resultPullback(v)
+        let (𝛁double1, 𝛁double2) = doublePullback(𝛁result1 + 𝛁result2)
+        return 𝛁double1 + 𝛁double2
+    }
+    return (value: result, pullback: pullback)
+}
+
+// Test.
+let x: Float = 3.0
+let (result, pullback) = fooValueWithPullback(x)
+print(result) // 36.0
+print(pullback(1)) // 24.0
+
+// Test the real `valueWithPullback` function.
+do {
+    let (result, pullback) = valueWithPullback(at: x, in: foo)
+    print(result) // 36.0
+    print(pullback(1)) // 24.0
+}
+```
+
+All other differentiation APIs are defined in terms of `valueWithPullback`.
+For example, here's how [`gradient(at:in:)`] is defined:
+
+```swift
+// `gradient` returns the partial derivative with respect to varying parameters for scalar-result
+// functions. It simply returns `pullback(1)`.
+func gradient<T, R>(
+    at x: T, in f: @differentiable (T) -> R
+) -> T.CotangentVector
+    where T : Differentiable, R : FloatingPoint & Differentiable,
+          R.CotangentVector == R {
+    let (value, pullback) = valueWithPullback(at: x, in: f)
+    return pullback(R(1))
+}
+
+print(gradient(at: x, in: foo)) // 24.0
+```
```

Comment:
Add a line below or above: 
"You can find [its implementation](https://github.com/apple/swift/blob/6e20cf1d2746435d879ccfeee9713aa7a32e1081/stdlib/public/core/AutoDiff.swift#L467) in the standard library."

Reply:
Adding:
```markdown
You can find [the implementation of
`gradient(at:in:)`](https://github.com/apple/swift/blob/6e20cf1d2746435d879ccfeee9713aa7a32e1081/stdlib/public/core/AutoDiff.swift#L467)
in the standard library.
```